coder/coder
# feat: add --experiments flag

# What this does
This PR makes the following changes:
- Deprecates the `--experimental` flag
- Adds a new flag `--experiments` which supports passing multiple comma-separated values or a wildcard value. 
- Exposes a new endpoint `/api/v2/experiments` that returns the list of enabled experiments.
- Deprecates the field `Features.Experimental` in favour of this new API.

# Why this does

Currently we only support setting a global boolean `--experimental`. For safe long-term feature development, we need to have a way to allow developers or users to conditionally enable or disable individual experimental code paths. This enables this use-case.

# How this does

- ~The cleanest way I found to implement the wildcard expansion behaviour is to use a type alias and implement the wildcard expansion in the `DeploymentConfig` reflection logic. I'm open to other possibilities.~ EDIT: This is now done on startup in `coderd.go`.
- The existing behaviour of setting `--experimental=true` is maintained for backward compatibility.
- Specifying the experiment wildcard will expand it to all experiments listed in `codersdk.ExperimentsAll`. Developers may wish to define additional experiments but _not_ include them in this set, and this is OK.
- Elected to use a simple `[]string` for this. I considered a `map[string]bool` but the performance gain for the current low value of N (<10) does not, to me, outweigh the complexity of the additional plumbing for Viper et al.

# Other things this does

- Added a fix from @mtojek for in the `apidocgen` relating to type aliases.
- Modified the `apitypings` script to support generating slice types. The existing code for maps appears to work fine here.
- Updated `develop.sh` to pass additional args after `--` to `$CODERD_SHIM`.

# How to test what this does

Run `./scripts/develop.sh -- --experiments='*'` and observe the local VSCode button on a running workspace.
Hit `/api/v2/experiments` and note the return value `['vscode_local']`.

You can specify nonexistent experiments and they will be accepted but have no effect. This will be important behaviour as we remove or promote experiments.

(Note: currently, the only experimental feature that exists is `vscode_local`. Others will come along in future.)
